### PR TITLE
refactor drip.drips

### PR DESCRIPTION
**Pull request by jiss2891**
## Description
* add lru_cache to `message_class_for` which is called every time we call `DripBase.send` method.
* add a couple of docstrings
* refactor configured_message_classes to make use of ChainMap (dict with steroids)

## Closes issue(s)
No issues associated.

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [x] Non-breaking change, refactor*.

*  Improves performance of current functionality and adds readability

### CODE REVIEW

**Review by kaozdl — DISMISSED** [review]

File: drip/drips.py [hunk 1]
```diff
@@ -21,24 +26,66 @@
...
+            'my_other_class': 'my.other.path.to.Drip',
+        }
+
+    A 'default' key is added when it's not present on DRIP_MESSAGE_CLASSES.
     """
-    conf_dict = getattr(settings, 'DRIP_MESSAGE_CLASSES', {})
-    if 'default' not in conf_dict:
-        conf_dict['default'] = 'drip.drips.DripMessage'
-    return conf_dict
+    default_config = {'default': 'drip.drips.DripMessage'}
```
Comment (nsantos16) [inline drip/drips.py:65]: Maybe add this as a constant outside of the function in order to make it more visible in the code

**Review by nsantos16 — DISMISSED** [review]
I leave a small suggestion, great work!

File: drip/drips.py (hunk 1 above)
Comment (jiss2891) [reply drip/drips.py:65]: Make a lot of sense, thanks Nico!

File: drip/drips.py [hunk 2]
```diff
@@ -20,25 +25,69 @@
...
+        }
+
+    The returning value of this method would be:
+
+        {
+            'default': 'drip.drips.DripMessage',
+            'my_other_class': 'my.other.path.to.DripMessage',
+        }
+
+    A 'default' key is added when it's not present on DRIP_MESSAGE_CLASSES.
```
Comment (brunomichetti) [inline drip/drips.py:65]: What happens if the users doesn't define a `default` class, and the `my_other_class` is `drip.drips.DripMessage`?
Is that case possible? Is it covered?

File: drip/drips.py [hunk 3]
```diff
@@ -20,25 +25,69 @@
...
-def message_class_for(name: str):
-    path = configured_message_classes()[name]
-    mod_name, klass_name = path.rsplit('.', 1)
-    mod = import_module(mod_name)
-    klass = getattr(mod, klass_name)
-    return klass
+@lru_cache()
+def message_class_for(name: str) -> Any:
+    """
+    Given a class' path, returns a reference to it.
```
Comment (brunomichetti) [inline drip/drips.py:75]:
```suggestion
    Given a class path, returns a reference to it.
```

File: drip/drips.py [hunk 4]
```diff
@@ -20,25 +25,69 @@
...
+    the ChainMap produced by configured_message_classes.
+    """
+    try:
+        path = configured_message_classes()[name]
+    except KeyError:
+        logging.error(
+            'Name "{}" not found in configured message classes.'.format(name)
+        )
+        raise MessageClassNotFound() from None
+    else:
```
Comment (brunomichetti) [inline drip/drips.py:86]: You don't need `else` here. If the exception is not raised, this block will be executed. Please remove the else

**Review by brunomichetti — CHANGES_REQUESTED** [review]

**Comment by brunomichetti** [comment]
Can this be tested? If it is please add some minor tests

File: drip/drips.py (hunk 2 above)
Comment (jiss2891) [reply drip/drips.py:65]:
```python
    # We set a default configuration dict.
    default_config = {'default': DEFAULT_DRIP_MESSAGE_CLASS}
    # Tries to get a user-defined config, falling back to an empty dict if config is not found.
    user_defined_config = getattr(settings, 'DRIP_MESSAGE_CLASSES', {})
    # Chain-map those dictionaries
    return ChainMap(user_defined_config, default_config)
```

File: drip/drips.py (hunk 4 above)
Comment (jiss2891) [reply drip/drips.py:86]: I believe it adds more readability to the code, in the sense that it's clear that the original developer wanted that code inside the `else` clause to be executed *only* when no errors occur. But it's just a personal preference. If the team prefers some approach to the other we can use that one.

**Comment by jiss2891** [comment]
> Can this be tested? If it is please add some minor tests

Sure thing!

File: drip/tests/test_drips.py [hunk 5]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+    def test_message_class_for(self):
+        default_message_classes_length = len(
+            configured_message_classes().items()
+        )
+
+        # adding a brand new MessageClass
+        setattr(
+            settings,
+            'DRIP_MESSAGE_CLASSES',
+            {'non-default-class': 'drip.drips.OtherDripClass'}
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:559]:
```suggestion
            {'non-default-class': 'drip.drips.OtherDripClass'},
```

File: drip/tests/test_drips.py [hunk 6]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+
+        message_classes = configured_message_classes()
+
+        self.assertEqual(
+            len(message_classes.items()),
+            default_message_classes_length + 1
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:566]:
```suggestion
            default_message_classes_length + 1,
```

File: drip/tests/test_drips.py [hunk 7]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+        self.assertEqual(
+            message_classes['default'],
+            DEFAULT_DRIP_MESSAGE_CLASS
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:570]:
```suggestion
            DEFAULT_DRIP_MESSAGE_CLASS,
```

File: drip/tests/test_drips.py [hunk 8]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+
+        # Replacing an existing Message Class
+        setattr(
+            settings,
+            'DRIP_MESSAGE_CLASSES',
+            {'default': 'drip.drips.OtherDripClass'}
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:577]:
```suggestion
            {'default': 'drip.drips.OtherDripClass'},
```

File: drip/tests/test_drips.py [hunk 9]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+
+        message_classes = configured_message_classes()
+
+        self.assertEqual(
+            len(message_classes.items()),
+            default_message_classes_length
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:584]:
```suggestion
            default_message_classes_length,
```

File: drip/tests/test_drips.py [hunk 10]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+        self.assertEqual(
+            message_classes['default'],
+            "drip.drips.OtherDripClass"
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:588]:
```suggestion
            'drip.drips.OtherDripClass',
```

File: drip/tests/test_drips.py [hunk 11]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+                'default': 'drip.drips.OtherDripClass',
+                'custom': 'custom.module.ClassName',
+            }
+        )
+
+        message_classes = configured_message_classes()
+
+        self.assertEqual(
+            len(message_classes.items()),
+            default_message_classes_length + 1
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:605]:
```suggestion
            default_message_classes_length + 1,
```

File: drip/tests/test_drips.py [hunk 12]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+        self.assertEqual(
+            message_classes['default'],
+            "drip.drips.OtherDripClass"
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:609]:
```suggestion
            'drip.drips.OtherDripClass',
```

File: drip/tests/test_drips.py [hunk 13]
```diff
@@ -542,6 +547,72 @@ def test_apply_or_queryset_ruletype(self):
...
+        )
+        self.assertEqual(
+            message_classes['custom'],
+            "custom.module.ClassName"
```
Comment (brunomichetti) [inline drip/tests/test_drips.py:613]:
```suggestion
            'custom.module.ClassName',
```

**Review by brunomichetti — CHANGES_REQUESTED** [review]
Great work! please update project version

**Pull request closed without merging**